src/vm/arm64/asmhelpers.S

```diff
@@ -376,6 +370,100 @@ LEAF_ENTRY JIT_PatchedCodeStart, _TEXT
     ret  lr
 LEAF_END JIT_PatchedCodeStart, _TEXT
 
+// void JIT_UpdateWriteBarrierState(bool skipEphemeralCheck)
+//
+// Update shadow copies of the various state info required for barrier
+//
+// State info is contained in a literal pool at the end of the function
+// Placed in text section so that it is close enough to use ldr literal and still
+// be relocatable. Eliminates need for PREPARE_EXTERNAL_VAR in hot code.
+//
+// Align and group state info together so it fits in a single cache line
+// and each entry can be written atomically
+//
+WRITE_BARRIER_ENTRY JIT_UpdateWriteBarrierState
+    PROLOG_SAVE_REG_PAIR_INDEXED   fp, lr, -16
+
+    // x0-x7 will contain intended new state
+    // x8 will preserve skipEphemeralCheck
+    // x12 will be used for pointers
+
+    mov x8, x0
+
+    PREPARE_EXTERNAL_VAR g_lowest_address, x12
+    ldr  x0, [x12]
+
+    PREPARE_EXTERNAL_VAR g_highest_address, x12
+    ldr  x1, [x12]
+
+#ifdef WRITE_BARRIER_CHECK
+    PREPARE_EXTERNAL_VAR g_GCShadow, x12
+    ldr  x2, [x12]
+#endif
+
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    PREPARE_EXTERNAL_VAR g_sw_ww_enabled_for_gc_heap, x12
+    ldrb w3, [x12]
+    cbz  x3, LOCAL_LABEL(SOFTWARE_WRITE_WATCH_DISABLED)
+    PREPARE_EXTERNAL_VAR g_sw_ww_table, x12
+    ldr  x3, [x12]
+LOCAL_LABEL(SOFTWARE_WRITE_WATCH_DISABLED):
+#endif
+
+    PREPARE_EXTERNAL_VAR g_ephemeral_low, x12
+    ldr  x4, [x12]
+
+    PREPARE_EXTERNAL_VAR g_ephemeral_high, x12
+    ldr  x5, [x12]
+
+    cbz  x8, LOCAL_LABEL(EphemeralCheckEnabled)
+    movz x4, #0
+    movn x5, #0
+LOCAL_LABEL(EphemeralCheckEnabled):
+
+    PREPARE_EXTERNAL_VAR g_card_table, x12
+    ldr  x6, [x12]
+
+#ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+    PREPARE_EXTERNAL_VAR g_card_bundle_table, x12
+    ldr  x7, [x12]
+#endif
+
+    // Update wbs state
+    adr  x12, LOCAL_LABEL(wbs_begin)
+    stp  x0, x1, [x12], 16
+    stp  x2, x3, [x12], 16
+    stp  x4, x5, [x12], 16
+    stp  x6, x7, [x12], 16
+
+    // Force updates to be visible other processors
+    dmb ishst
```

Comment:
I added the barrier to order GC parameter changes before memory writes which logically occur after GC parameter changes.
+ heap is resized to allocate sufficient space to allocate an new object
+ ephemeral heap is changed

For instance the `JIT_CheckedWriteBarrier` range must be visible to other processors before an object which is allocated in the new resized heap.  This barrier guarantees that store ordering.

Likewise if the ephemeral size is changed, other processors must see the new parameters before an object is created in the new ephemeral range.

I was debating suspending the runtime, but decided on the memory barrier instead.